ENH: implement `spatial_shuffle` method

martinfleis opened the pull request:
This supersedes #104 and closes #88.

~I still need to add tests.~

I am not allowing custom Series in `by` as in that case you would just use `set_index` directly.

tastatham commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,101 @@ def geohash(self, string=True, p=12):
+        by="hilbert",
+        p=None,
+        calculate_partitions=True,
+        drop_index=True,
```
[drop_index](https://docs.dask.org/en/latest/_modules/dask/dataframe/core.html#DataFrame.set_index) is not necessary because we are not using a column to partition the Dask GeoDataFrame - so there is nothing to drop.

martinfleis replied:
This controls whether the original index is dropped or not. If you have some meaningful info there, you may want to keep it. If it's just a range, you can drop it. See how it is passed below to `set_index`.

jorisvandenbossche replied:
We might still mention that in eg the notes section that this will remove any existing index? 

Do we still want to option to not _set_ the distance as the index? Or you can do a manual `.reset_index(drop=True)` in case you don't want it? 
Eg when writing the result of `spatial_shuffle` to a partitioned (parquet) dataset, you might not necessarily care about the exact distance values for later analyses, and then it just occupies extra space in the files.

martinfleis replied:
> We might still mention eg in the notes section that this will remove any existing index?

It is on the line 452.

> Do we still want to option to not set the distance as the index? Or you can do a manual `.reset_index(drop=True)` in case you don't want it?

I would vote for the latter. If we added the keyword, we would in the end only used it to call `.reset_index(drop=True)` under the hood.

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+        Shuffle the data into spatially consistent partitions.
+
+        This realigns the dataset to be spatially sorted, i.e. geometries that are
+        spatially near each other together will be within the same partition. This is
```
```suggestion
        spatially near each other will be within the same partition. This is
```

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+        Parameters
+        ----------
+        by : string (default 'hilbert')
+            sorting method, one of {'hilbert', 'morton, 'geohash'}. See
```
```suggestion
            Spatial sorting method, one of {'hilbert', 'morton', 'geohash'}. See
```

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+            ``hilbert_distance``, ``morton_distance`` and ``geohash`` methods for
+            details.
+        p : int (default None)
+            precision of the sorting method. Defaults to 15 for Hilbert and Morton
```
```suggestion
            Precision of the sorting method. Defaults to 15 for Hilbert and Morton
```

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+            precision of the sorting method. Defaults to 15 for Hilbert and Morton
+            curves and 12 for geohash.
+        calculate_partitions : bool (default True)
+            calculate new spatial partitions after shuffling
```
```suggestion
            Calculate new spatial partitions after shuffling
```

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+            match the values returned by the sorting method.
+
+        **kwargs
+            keyword arguments passed to ``set_index``.
```
```suggestion
            Keyword arguments passed to ``set_index``.
```

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+            shuffle = kwargs.pop("shuffle")
+            if shuffle != "tasks":
+                warnings.warn(
+                    "Temporarily forcing the 'tasks' shuffle mode due to an upstream issue."
```
Do we have an issue about this?

jorisvandenbossche replied:
Ah, this is https://github.com/geopandas/dask-geopandas/issues/59

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+            sorting method, one of {'hilbert', 'morton, 'geohash'}. See
+            ``hilbert_distance``, ``morton_distance`` and ``geohash`` methods for
+            details.
+        p : int (default None)
```
Not for this PR, but I am wondering if we shouldn't rename this for all those methods to `precision` for clarity

martinfleis replied:
That is a good point.

jorisvandenbossche commented on dask_geopandas/core.py:
```diff
@@ -429,6 +432,100 @@ def geohash(self, string=True, p=12):
+        -----
+        This method, similarly to ``calculate_spatial_partitions``, is computed
+        partially eagerly as it needs to load the active geometry column and compute
+        spatial partitions.
```
This is the case when `calculate_spatial_partitions=True`? If so, maybe mention in the above paragraph you can make it fully lazily by passing that keyword.  

And for example, I suppose that if you do a directly following write to disk, such as `ddf.spatial_shuffle().to_parquet()`, you actually want to turn this off? Because then dask can combine all things in a single graph, and we will calculate the bounds of each partition anyway when writing to Parquet.

martinfleis replied:
> This is the case when calculate_spatial_partitions=True? If so, maybe mention in the above paragraph you can make it fully lazily by passing that keyword.

I don't think so, it seems to be computed eagerly no matter the `calculate_spatial_partitions` option because dask needs to know the divisions for the new index. And to get those, it needs to load geometry and compute Hilbert curve.

jorisvandenbossche replied:
Yeah, the `set_index` documentation (https://docs.dask.org/en/latest/generated/dask.dataframe.DataFrame.set_index.html) has this about the `compute` keyword:

> Whether or not to trigger an immediate computation. Defaults to False. Note, that even if you set `compute=False`, an immediate computation will still be triggered if `divisions` is `None`.

jorisvandenbossche replied:
Maybe we can then clarify the note a little bit? Eg "... as it needs to calculate the distances for all existing partitions before it can determine the divisions for the new spatially-shuffled partitions"

jorisvandenbossche reviewed (commented):
Looking good!

martinfleis commented:
I still have to debug the CI failures but they seem to be related to the geohash implementation, not `spatial_shuffle`.

martinfleis commented:
I have fixed the dtype issue in `geohash` affecting Windows envs and added tests for all sorting methods using the real-world data. 

The final failure caused by this PR is in `37-latest` and is causes by incorrect sorting of dask dataframe, that was fixed in dask 2021.03.0. So I'd skip that for now.

martinfleis commented:
CI is as green as it gets right now. Consider this PR ready.

jorisvandenbossche commented:
> remove p from goehash (unused)

That's because it is only used when being converted to a string value?

martinfleis commented:
> That's because it is only used when being converted to a string value?

Yes, which is not used here.

jorisvandenbossche reviewed (commented):
Looks good!

I merged the set_index fix, so you should be able to clean that up here before merging

jorisvandenbossche commented:
> move to GeoDataFrame only 

It doesn't work for GeoSeries out of the box? (it's certainly less useful / less typical to do it with a GeoSeries though)

martinfleis commented:
> It doesn't work for GeoSeries out of the box? (it's certainly less useful / less typical to do it with a GeoSeries though)

GeoSeries / Series does not have `set_index` in Dask apparently.

jorisvandenbossche commented:
Thanks!
